runtime/oti/ObjectMonitor.hpp

```diff
@@ -473,6 +481,115 @@ class VM_ObjectMonitor
 		recordMonitorExit(currentThread, object, &currentThread->jniMonitorEnterRecords);
 	}
 
+	/**
+	 * Increment reservedCounter stored in the J9Class.
+	 *
+	 * @param clazz[in] the object's J9Class
+	 */
+	static VMINLINE void
+	incrementReservedCounter(J9Class *clazz)
+	{
+		/*
+		 * reservedCounter is a per class counter that tracks lockword state transitions to Reserved.
+		 * It is later used to help determine what the initial lockword for a created object should be.
+		 */
+		U_16 reservedCounter = clazz->reservedCounter;
+
+		/* If incrementing the reservedCounter would cause an overflow, first divide cancelCounter and reservedCounter by 2. */
+		if (reservedCounter < (U_16)0xFFFF) {
+			clazz->reservedCounter = reservedCounter + 1;
+		} else {
+			/*
+			 * cancelCounter is a per class counter that tracks lockword state transitions to Flat.
+			 * It is later used to help determine what the initial lockword for a created object should be.
+			 */
+			U_16 cancelCounter = clazz->cancelCounter;
+
+			clazz->reservedCounter = (reservedCounter >> 1) + 1;
+			clazz->cancelCounter = (cancelCounter >> 1);
+		}
```

Comment:
I added a commit to remove `OBJECT_HEADER_LOCK_V2_FIRST_RECURSION_BIT` and `OBJECT_HEADER_LOCK_V2_RECURSION_MASK`.

Can you clarify what you are referring to by "the instance"?

There is no locking or anything to prevent concurrent updates to the class counters. So updates by one thread may get clobbered by updates from other threads. It was decided that this would be okay since the values of the counters don't need to keep perfect track of the reservation and cancellation events.

Reply:
My point is that two threads might lock distinct objects (instances) of the same class. How do we not have a race with respect to classes?

Reply:
@keithc-ca we have to live with that race condition.  a conscious trade-off decision, that is. no correctness issue, but collected data driving the appropriate optimization.

Reply:
I think that's something worth mentioning in `ObjectLockword.md`.

Reply:
I've added a mention to it in the documentation.